### PR TITLE
Allow for varying an invocations fireDate by a specified number of seconds

### PR DESCRIPTION
**Pull request by andy-g**
I required functionality to randomly vary my invocations by a specified range, for example, specifying a schedule for every hour, but allowing a variation of 5 minutes, in which case the 12:00 Invocation may run at anytime between 11:55 - 12:05.

Although this is relatively specific functionality I assume that it may be useful to others, so decided to create this Pull Request.

All existing functionality remains unchanged, but a new `variation` property is available on the rule object, which if specified will adjust the resulting invocations accordingly.

```javascript
var rule = new schedule.RecurrenceRule();
rule.minute = 0;
rule.variation = 300;

var j = schedule.scheduleJob(rule, function(){
    console.log('within 5 minutes before or after the hour');
});
```

I've also added some associated tests.

### CODE REVIEW

File: lib/schedule.js [hunk 1]
```diff
@@ -213,14 +216,24 @@ var DoesntRecur = new RecurrenceRule();
...
 
   this.timerID = null;
+
+  if (this.recurrenceRule.variation && this.recurrenceRule.variation !== 0) {
+    var variation = Math.floor(this.random() * this.recurrenceRule.variation * 2) + (this.recurrenceRule.variation * -1);
+    this.fireDate.setTime(this.fireDate.getTime() + (variation * 1000));
+  }
 }
 
+Invocation.prototype.random = function() {
```
Comment (santigimeno) [inline lib/schedule.js]: Why not use Math.random directly?

File: test/convenience-method-test.js [hunk 2]
```diff
@@ -167,6 +167,25 @@ module.exports = {
           }, 1000);
         }*/
   },
+  ".scheduleJob({...}, fn)": {
+    "Runs job at interval based on object, repeating indefinitely, with 5 second variation": function(test) {
+      test.expect(4);
```
Comment (santigimeno) [inline test/convenience-method-test.js]: Not 100% sure, but couldn't this job fire 5 times sometimes?

File: test/recurrence-rule-test.js [hunk 3]
```diff
@@ -255,6 +255,60 @@ module.exports = {
...
+      test.done();
+    },
+    "specify 5 second variation (override variation to test exact fireDate)": function(test) {
+      test.expect(8);
+
+      var rule = new schedule.RecurrenceRule();
+      rule.second = [0,15,30,45];
+      rule.variation = 5;
+
+      schedule.Invocation.prototype.random = function(){
```
Comment (santigimeno) [inline test/recurrence-rule-test.js]: Oh, I see the reason of `Invocation.prototype.random` now. Couldn't we just mock Math.random?

**Comment by santigimeno** [comment]
Thanks for the PR. I left some comments and questions. Apart from that it looks fine to me. I agree with you that maybe it's a little too specific. @tejasmanohar @jonhester what do you think?

File: test/recurrence-rule-test.js (hunk 3 above)
Comment (andy-g) [inline test/recurrence-rule-test.js]: We could and I originally did, but I wasn't sure what other compilations may result from mocking Math.random. If you think it'd be cleaner I'd be happy to revert to a Math.random mock and remove the `Invocation.prototype.random` implementation.

File: lib/schedule.js (hunk 1 above)
Comment (tejasmanohar) [inline lib/schedule.js]: same question as @santigimeno

File: test/convenience-method-test.js (hunk 2 above)
Comment (andy-g) [inline test/convenience-method-test.js]: No, only the invocations for 15,30,45 & 60 seconds will run and the last one will be between 55 & 65 seconds into the run, the 5th invocation would only run between 70 & 80 seconds into the run. The 65250 timeout should limit it to 4 invocations.

**Comment by tejasmanohar** [comment]
Honestly, I personally am not convinced that it's going to serve the general audience of this module. Seems like you have a really specific use case in mind. Probably better to stay as a fork.

Thoughts, @jonhester?

**Comment by andy-g** [comment]
I'm currently using this for an IoT implementation on my RaspberryPi, it allows lights and devices to be scheduled with a specified deviation. I realise that it's a specific use case, but I do feel that it can be very useful, would be used by others, and in light of it not affecting existing functionality I do feel it may be worth including. Totally respect your call though.

**Comment by tejasmanohar** [comment]
I personally don't see this functionality in other job schedulers or think it's a fit for the module. Would like to keep it as lean as possible covering most use cases.

That said, glad to hear you're using node-schedule and thanks for the PR!

**Comment by tejasmanohar** [comment]
By no means is this locked though- If many people comment on this thread expressing that this is something they're struggling with etc and want in node-schedule, I'd be in. In the end, it's the community's choice when voiced :smile:

**Comment by gabek** [comment]
I was just looking for this functionality :)

**Comment by tejasmanohar** [comment]
I still think it's too specific... people may want different ways to implement _variation_ for example. Plus, I don't know another job scheduler that has anything similar-- seems like an application-level concern. That said, if you have a generic suggestion for how this module could better support use cases like this, please open a new issue (or PR) with a detailed proposal.

**Pull request closed without merging**